### PR TITLE
sscs 9403 added form ref and benefit types

### PR DESCRIPTION
**Pull request by chrisdavecm**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/SSCS-9403

### Change description ###
Adding new form sscs1u

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -212,7 +212,7 @@ dependencies {
   compile group: 'com.github.hmcts', name: 'java-logging', version: versions.reformLogging
   compile group: 'uk.gov.hmcts.reform', name: 'service-auth-provider-client', version: '4.0.0'
 
-  compile group: 'com.github.hmcts', name: 'sscs-common', version: '4.4.45'
+  compile group: 'com.github.hmcts', name: 'sscs-common', version: '4.4.47.RC-SSCS-9403'
```
Comment (gmikerlemos) [inline build.gradle:215]: merging RC lib into master

**Review by gmikerlemos — COMMENTED** [review]
Not sure about the dependency lib pointing to a RC, If the team is fine with it good to go.

File: build.gradle (hunk 1 above)
Comment (chrisdavecm) [reply build.gradle:215]: can't release it till its tested, will be released when deployed

File: src/functionalTest/resources/exception/all_fields_entered_uc.json [hunk 2]
```diff
@@ -3,7 +3,7 @@
   "case_type_id": "case_type_id",
   "po_box": "po_box",
   "po_box_jurisdiction": "bulkscan",
-  "form_type": "SSCS1",
+  "form_type": "SSCS1PEU",
```
Comment (jackmaloney) [inline src/functionalTest/resources/exception/all_fields_entered_uc.json:6]: Do you mind adding an equivalent test for the new SSCS1U form with the new fields?

File: src/main/java/uk/gov/hmcts/reform/sscs/transformers/SscsCaseTransformer.java [hunk 3]
```diff
@@ -254,18 +270,39 @@ private BenefitType getBenefitType(Map<String, Object> pairs) {
...
+
+        if (!validProvidedBooleanValues.isEmpty()
+            && !isExactlyZeroBooleanTrue(pairs, errors, validProvidedBooleanValues.toArray(new String[validProvidedBooleanValues.size()]))) {
             // If one is set to true, extract the string indicator value (eg. IS_BENEFIT_TYPE_PIP) and lookup the Benefit type.
             if (isExactlyOneBooleanTrue(pairs, errors, validProvidedBooleanValues.toArray(new String[validProvidedBooleanValues.size()]))) {
                 String valueIndicatorWithTrueValue = validProvidedBooleanValues.stream().filter(value -> extractBooleanValue(pairs, errors, value)).findFirst().orElse(null);
-                Optional<Benefit> benefit = BenefitTypeIndicator.findByIndicatorString(valueIndicatorWithTrueValue);
-                if (benefit.isPresent()) {
-                    code = benefit.get().name();
+                if ("is_benefit_type_other".equals(valueIndicatorWithTrueValue)) {
```
Comment (jackmaloney) [inline src/main/java/uk/gov/hmcts/reform/sscs/transformers/SscsCaseTransformer.java:279]: Can we add this variable to SscsConstants.java?

File: src/main/java/uk/gov/hmcts/reform/sscs/transformers/SscsCaseTransformer.java [hunk 4]
```diff
@@ -254,18 +270,39 @@ private BenefitType getBenefitType(Map<String, Object> pairs) {
...
+                    if (!StringUtils.isEmpty(benefitTypeOther)) {
+                        Optional<Benefit> benefit = Benefit.findBenefitByDescription(benefitTypeOther);
+                        if (benefit.isPresent()) {
+                            code = benefit.get().getShortName();
+                            return BenefitType.builder().code(code).description(benefit.get().getDescription()).build();
+                        } else {
+                            errors.add("enter valid benefit type in 'benefit_type_other");
+                        }
+                    } else {
+                        errors.add("benefit_type_other field is empty");
```
Comment (jackmaloney) [inline src/main/java/uk/gov/hmcts/reform/sscs/transformers/SscsCaseTransformer.java:291]: And use this variable here (and error message above)

**Review by jackmaloney — APPROVED** [review]

File: src/test/java/uk/gov/hmcts/reform/sscs/transformers/SscsCaseTransformerTest.java [hunk 5]
```diff
@@ -208,10 +233,34 @@ public void givenBenefitTypePipWithIsOtherBenefit_thenErrorMessage() {
         pairs.put(BenefitTypeIndicatorSscs1U.UC.getIndicatorString(), false);
         pairs.put(BENEFIT_TYPE_OTHER, "Attendance Allowance");
         CaseResponse result = transformer.transformExceptionRecord(sscs1UExceptionRecord, false);
-        assertFalse(result.getErrors().isEmpty());
+        assertTrue(result.getErrors().size() == 1);
         assertEquals("is_benefit_type_pip and benefit_type_other have contradicting values", result.getErrors().get(0));
     }
 
+    @Test
+    public void givenBenefitTypePipWithIsOtherBenefitYes_thenErrorMessage() {
```
Comment (jackmaloney) [inline src/test/java/uk/gov/hmcts/reform/sscs/transformers/SscsCaseTransformerTest.java:249]: Not sure this test name makes sense. Mentions PIP but the test has nothing to do with PIP - can you update?

File: src/test/java/uk/gov/hmcts/reform/sscs/transformers/SscsCaseTransformerTest.java [hunk 6]
```diff
@@ -208,10 +233,34 @@ public void givenBenefitTypePipWithIsOtherBenefit_thenErrorMessage() {
...
+        pairs.put(IS_BENEFIT_TYPE_OTHER, "Yes");
+        pairs.put(BenefitTypeIndicatorSscs1U.ESA.getIndicatorString(), "Yes");
+        pairs.put(BenefitTypeIndicatorSscs1U.UC.getIndicatorString(), "Yes");
+        CaseResponse result = transformer.transformExceptionRecord(sscs1UExceptionRecord, false);
+        assertTrue(result.getErrors().size() == 1);
+        assertEquals("is_benefit_type_esa, is_benefit_type_uc and is_benefit_type_other have contradicting values", result.getErrors().get(0));
+    }
+
+    @Test
+    public void givenBenefitTypePipWithIsOtherBenefitTrue_thenErrorMessage() {
```
Comment (jackmaloney) [inline src/test/java/uk/gov/hmcts/reform/sscs/transformers/SscsCaseTransformerTest.java:261]: Not sure this test name makes sense. Mentions PIP but the test has nothing to do with PIP - can you update?

File: src/main/java/uk/gov/hmcts/reform/sscs/constants/WarningMessage.java [hunk 7]
```diff
@@ -35,6 +35,7 @@
     APPOINTEE_POSTCODE(ADDRESS_POSTCODE, "Appointee postcode"),
     REPRESENTATIVE_POSTCODE(ADDRESS_POSTCODE, "Representative postcode"),
     BENEFIT_TYPE_DESCRIPTION(SscsConstants.BENEFIT_TYPE_DESCRIPTION, "Benefit type description"),
+    BENEFIT_TYPE_OTHER(SscsConstants.BENEFIT_TYPE_OTHER, "Benefit type other"),
```
Comment (jackmaloney) [inline src/main/java/uk/gov/hmcts/reform/sscs/constants/WarningMessage.java:38]: Do you need this? At least I don't think '"Benefit type other"' can ever be shown as it doesn't make sense in SSCS world which is where this error gets shown. If it is being shown, we probably need to change the wording...

File: src/integrationTest/java/uk/gov/hmcts/reform/sscs/controllers/OcrValidationTest.java [hunk 8]
```diff
@@ -133,7 +133,7 @@ public void should_return_200_with_status_warnings_when_ocr_form_validation_requ
     public void should_return_200_when_ocr_form_with_form_type_sscs1peu_and_hearing_sub_type_validation_request_data_are_empty() throws Throwable {
         when(authTokenValidator.getServiceName(SERVICE_AUTH_TOKEN)).thenReturn("test_service");
 
-        String content = readResource("mappings/ocr-validation/invalid-ocr-data-with-hearing-sub-type.json");
+        String content = readResource("mappings/ocr-validation/invalid-ocr-data-with-hearing-sub-type-sscs1peu.json");
```
Comment (jackmaloney) [inline src/integrationTest/java/uk/gov/hmcts/reform/sscs/controllers/OcrValidationTest.java:136]: Can we add a test for validating the new sscs1u form?

File: src/integrationTest/java/uk/gov/hmcts/reform/sscs/controllers/SscsBulkScanExceptionRecordCallback.java [hunk 9]
```diff
@@ -256,7 +257,7 @@ public void auto_scan_should_handle_callback_and_return_caseid_and_state_case_cr
 
         when(authTokenValidator.getServiceName(SERVICE_AUTH_TOKEN)).thenReturn("test_service");
 
-        HttpEntity<ExceptionRecord> request = new HttpEntity<>(autoExceptionCaseData(caseDataWithMrnDate(MRN_DATE_YESTERDAY_DD_MM_YYYY, this::addAppellant)),
+        HttpEntity<ExceptionRecord> request = new HttpEntity<>(autoExceptionCaseData(caseDataWithMrnDate(MRN_DATE_YESTERDAY_DD_MM_YYYY, this::addAppellant, "SSCS1PEU")),
```
Comment (jackmaloney) [inline src/integrationTest/java/uk/gov/hmcts/reform/sscs/controllers/SscsBulkScanExceptionRecordCallback.java:260]: Can we add an equivalent test to cover the new SSCS1U form?

**Review by jackmaloney — COMMENTED** [review]
Also, there is no test in src/test/java/uk/gov/hmcts/reform/sscs/validators/SscsCaseValidatorTest.java to cover the new scenario you have written which swaps out the field it looks at depending on form type. Mind adding?

**Review by jackmaloney — APPROVED** [review]

**Pull request merged**